### PR TITLE
Release

### PR DESCRIPTION
**Pull request by Kaaserne**
Note: for the magic numbers please refer to the pipeline `mlx90640_processor_c::get_temperature_pixel` for the meaning of these numbers. Each class is responsible for their own calculations and each section has its own class.

### CODE REVIEW

File: code/headers/mlx90640.hpp [hunk 1]
```diff
@@ -22,11 +20,11 @@ namespace r2d2::thermal_camera {
         mlx90640_processor_c mlx_processor;
 
         /**
-         * Changes nth bit to 'to' in source.
+         * Changes nth bit to 'to' in 'source'.
          *
          * @param int source. The to be changed value.
```
Comment (LRstudentHU) [inline code/headers/mlx90640.hpp:25]: `int` -> `uint16_t`

File: code/src/data_extractor.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        extracted_data = apply_treshold(extracted_data, exceeds, minus);
+        return extracted_data;
+    }
+
+    int data_extractor_s::extract_data(const int value, const uint16_t and_bits,
+                                       const uint8_t shifted) {
+        return (value & and_bits) >> shifted;
+    }
+
+    void data_extractor_s::check_within_limits(int &row, int &col) {
```
Comment (LRstudentHU) [inline code/src/data_extractor.cpp:28]: You could use `std::clamp` instead of this function.

**Review by LRstudentHU — CHANGES_REQUESTED** [review]
Looks very good. Can't say anything about the calculations themselves.

File: code/src/data_extractor.cpp (hunk 2 above)
Comment (Kaaserne) [reply code/src/data_extractor.cpp:28]: Cplusplusreference says it is defined in `<algorithm>` but it cannot find the function specified.

File: code/headers/alpha_comp_extractor.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+     */
+    class alpha_comp_extractor_c : public mlx_extractor_c {
+    private:
+        int row, col;
+
+    public:
+        alpha_comp_extractor_c(mlx90640_i2c_c &bus, int row, int col);
+
+        void extract(mlx_parameters_s &params) override;
+    };
```
Comment (nicovanbentum) [inline code/headers/alpha_comp_extractor.hpp:17]: missing docs for member functions

File: code/headers/cp_compensator.hpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+     */
+    class cp_compensator_c : public mlx_extractor_c {
+    private:
+        reading_pattern pattern;
+
+    public:
+        cp_compensator_c(mlx90640_i2c_c &bus, const reading_pattern pattern);
+
+        void extract(mlx_parameters_s &params) override;
+    };
```
Comment (nicovanbentum) [inline code/headers/cp_compensator.hpp:15]: missing docs for member functions

File: code/headers/data_extractor.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <cstdint>
+
+namespace r2d2::thermal_camera {
+    class data_extractor_s {
+    public:
+        // static class
+        data_extractor_s() = delete;
+
```
Comment (nicovanbentum) [inline code/headers/data_extractor.hpp:9]: Why is this implemented as a deleted class and not just a namespace with functions? (docs would help)

File: code/headers/kv_coefficient.hpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+#include <mlx_extractor.hpp>
+
+namespace r2d2::thermal_camera {
+    /**
+     * Sets the Kv coeffecient
+     *
+     */
+    class kv_coefficient_c : public mlx_extractor_c {
+    private:
+        int row, col;
```
Comment (nicovanbentum) [inline code/headers/kv_coefficient.hpp:12]: can the rows or columns be negative values? if not these should be unsigned.

File: code/headers/mlx90640_processor.hpp [hunk 7]
```diff
@@ -1,94 +1,59 @@
...
-         * See apply_treshold and extract_data functions for the parameter info.
+         * Gets the VDD sensor parameters. Common for all pixels.
          *
-         * @return int - the processed data.
+         * @return float
          */
-        int get_compensated_data(const registers reg_addr,
-                                 const uint16_t and_bits, const uint8_t shifted,
-                                 const uint16_t exceeds, const int minus) const;
+        void set_Vdd();
```
Comment (nicovanbentum) [inline code/headers/mlx90640_processor.hpp:41]: docs doesn't match function

File: code/headers/ta_extractor.hpp [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#include <mlx_extractor.hpp>
+
+#pragma once
+
```
Comment (nicovanbentum) [inline code/headers/ta_extractor.hpp:4]: wrong order of pragma and include

**Review by nicovanbentum — CHANGES_REQUESTED** [review]
Please check again for missing documentation.

**Pull request closed without merging**